### DISCUSSION TITLE
Should casting to a generic type type-check?

### DISCUSSION THREAD

**Discussion by alxrcs** [microsoft/pyright#4217] [Q&A] [answered]
Assume the following example:

```python
from typing import TypeVar, Type, cast, Generic
from copy import copy

T = TypeVar("T", bound="Castable")

class Castable(Generic[T]):
    t: Type[T]

    def to_t(self) -> T:
        c = copy(self)
        c.__class__ = self.t
        return cast(T, c)

    def __init__(self, t: Type[T]):
        self.t = t
```

pyright 1.1.280 shows a somewhat confusing error of `Type "T@Castable" cannot be assigned to type "T@Castable"`
mypy 0.991 doesn't complain about this.

This should typecheck, right? Or is there something I'm missing and it shouldn't?

**Comment by erictraut** [comment, marked as answer]
This should work without error. I've created a [bug report](https://github.com/microsoft/pyright/issues/4218).
